Remove Couch backend (part 1)

millerdev opened the pull request:
## Summary
A new attempt at removing the code that uses the Couch forms & cases backend, from the edges this time. Most references to `should_use_sql_backend` and other Couch backend specific code has been removed from the `change_feed`, `cleanup`, and `data_interfaces` apps. This mainly removes tests, but some non-test code has also been removed where it appears to be safe.

Notably the `change_feed` app still contains code that references Couch form and case feeds since it was not clear if it is safe to remove them yet. Datadog showed some activity on those feeds in the past week, which is unexpected and needs further investigation.

:tropical_fish: Review by commit.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

No new tests added. Tests have been modified, and a few have been moved.

### QA Plan

No QA.

### Safety story

None of the removed code should be in use since all domains have been migrated to SQL.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Remove docs for should_use_sql_backend

millerdev committed:
Couch backend out of corehq/apps/change_feed

millerdev committed:
Couch backend out of corehq/apps/cleanup

millerdev committed:
Couch backend out of corehq/apps/data_interfaces

calellowitz commented on corehq/apps/data_interfaces/tests/test_scheduling_integration.py:
```diff
@@ -1349,6 +1319,7 @@ def test_next_available_monthly_slot(self, utcnow_patch, send_patch):
             self.assertTrue(instance.active)
 
 
+@use_sql_backend
```
Do we still need this decorator?

snopoke replied:
I think we do since the form processing code still defaults to couch without this.

calellowitz commented on corehq/apps/change_feed/tests/test_data_sources.py:
```diff
@@ -161,36 +151,6 @@ def location_data():
 
 
-@contextmanager
-def stock_data():
```
Is there a sql equivalent of this somewhere?

snopoke replied:
Yes, it's called `ledger_data`

calellowitz reviewed (approved):
Couple small questions but I could not be more excited about this! Kills couch code and may speed up the build, what more could we want? Do we need to worry about any old local hosters that still have couch domains, that we would need to alert about the need to migrate now that it is no longer optional?

snopoke commented on docs/forms_and_cases.rst:
```diff
@@ -127,17 +127,6 @@ For more details see:
 * :code:`corehq.form_processor.interfaces.dbaccessors.LedgerAccessors`
 
 
-Branching
```
If there are still other usages of this function then we should keep the docs around.

millerdev replied:
My thought was that this function is now deprecated and no new code should use it. This PR is the beginning of a multi-step process to remove all references to it. I guess I could wait to remove the docs until removing the last reference. Do you think that would be better?

snopoke replied:
No, you're right - we shouldn't need this any more.

snopoke reviewed (approved):
nothing blocking

millerdev commented:
> Do we need to worry about any old local hosters that still have couch domains, that we would need to alert about the need to migrate now that it is no longer optional?

[They were warned](https://github.com/dimagi/commcare-cloud/blob/master/docs/changelog/0044-migrate-forms-to-sql.md) and should have run the migration weeks ago if necessary.

millerdev committed:
isort

millerdev committed:
create_case still needs should_use_sql_backend

This commit moves create_case to a more logical place so all references to should_use_sql_backend can be removed from the data_interfaces app.

millerdev committed:
Fix test for USE_PARTITIONED_DATABASE=yes

Not sure how this test was ever passing in the past with USE_PARTITIONED_DATABASE=yes and a multi-shard config since AutomaticUpdateRule.iter_cases is called for each shard. Returning a list on a multi-shard config means the same case would be processed for each shard db. Returning an iterator causes it to be exhausted on the first shard, and all other shards to yield no cases.

snopoke commented on corehq/apps/data_interfaces/tests/test_auto_case_updates.py:
```diff
@@ -1031,7 +1031,7 @@ def test_scheduled_task_run(self):
         with _with_case(self.domain, 'person', datetime.utcnow()) as case:
             with patch('corehq.apps.data_interfaces.models.AutomaticUpdateRule.iter_cases') as iter_cases_patch:
-                iter_cases_patch.return_value = [case]
+                iter_cases_patch.return_value = iter([case])
```
Do you think this is the best way to mock this? Or would it be better to mock it with an implementation that only returns cases for one of the partitions?

This way is simpler and does work but it's not very clear what's happening. Just curious on your opinion.

millerdev replied:
This was the smallest change I could think of to make the test pass while retaining what it was intended to do. Mocking with something that is aware of db partitions might make it easier to know what's going on if something failed related to partitions, but has the disadvantage of having partition details hard-coded into the test. This test should not have to care about partitions since that is a different abstraction layer.

I actually like this solution since it works the same on single-db and multi-shard configurations.

snopoke replied:
:+1: perhaps adding a comment would be useful for anyone else who comes across this

millerdev replied:
The change was reverted in https://github.com/dimagi/commcare-hq/pull/30270/commits/ff25a308f252f61e235381c6080cbae26e2cd279. These tests have not historically been running in a sharded configuration. See also https://github.com/dimagi/commcare-hq/pull/30279.

millerdev committed:
Remove unused import

millerdev committed:
Do not use partitioned SQL database

Unlike @run_with_all_backends, tests that use @use_sql_backend are run with a _sharded_ SQL backend. Many tests will not work with a sharded backend.

millerdev committed:
Revert "Fix test for USE_PARTITIONED_DATABASE=yes"

This reverts commit 0e8f8158bd6f9d822734b49070fdfab521849f5c.

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/no-couch-backend-part1

mkangia commented:
:1st_place_medal:
